### ISSUE TITLE
Can we strengthen the traceful / traceful+option bind?

### ISSUE THREAD

**Issue by cwaldm** [REPROSEC/dolev-yao-star-extrinsic#90]
I ran into a problem with `let*?` where I couldn't show an assert, that I think should be true and showable.
Here is a minimal example:
```
let*? x = return (Some 0) in
assert(x = 0);
```
The assert can not be shown currently.

I looked at the comment for `let?`, which says that in the following similar example
```
let f = Some 9 in
let? x = f in
assert(x = 9);
Some ()
```
the assert can only be shown, if we change the type of `let?` to:
```
val (let?): #a:Type -> #b:Type -> x:option a -> (y:a{x == Some y} -> (option b)) -> option b
```
so that the second function of the bind has more information on its first argument, i.e., it can assume that it is only called for `Some`s.

Then, I tried to apply the same trick to `let*?`:
letting the second argument of bind know, that its first argument will be a `Some`.
Something like this:
```
 f:( (y:a{fst (x tr0) == Some y}) -> (traceful (option b)) )
```
However, the overall arguments of `let*?` are in the wrong order to state that property,
i.e., the trace `tr0` above is supposed to be the trace at the very end of the bind.

I can define a new bind with the orders swapped:
```
val tmp: #a:Type -> #b:Type -> x:traceful (option a) -> tr0:trace ->  f:( (y:a{fst (x tr0) == Some y}) -> (traceful (option b)) ) -> (out:(option b & trace){tr0 <$ snd out})
```
And re-writing the `let*?` sugar explicitly using this new bind, I can indeed show the above assert:
```
  let f = return (Some 0) in
  let g = return (Some "a") in
  let* tr = get_trace in
  let x = tmp f tr (fun x -> 
    assert(x = 0);
    g
  ) in
  ...
```

But if I use this to define `let*?`, e.g., like `let (let*?) #a #b x f tr0 = tmp #a #b x tr0 f` we loose the added knowledge again. That is, the very first example, still doesn't work.

Does someone have an idea how to resolve this?
I.e., get the behavior of the `tmp` bind but with the convenient `let*?` and `;*?` syntax.

**Comment by TWal** [comment]
I agree this is unfortunate, I have been trying to do that for `let*?` when implementing the trick for `let?` but couldn't for the same reason you have found. But it might be the case that with enough hacking we could get this to work, I don't know.
